MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java

```diff
@@ -0,0 +1,1080 @@
+        // once we've determined that scenarios occur, we loop through the ones that we generated
+        // and use the random force to drive opfor generation (#required lances multiplies the BV budget of all 
+        for(int scenarioIndex = 0; scenarioIndex < track.getRequiredLanceCount(); scenarioIndex++) {
+            // if we haven't already used all the player forces and are required to randomly generate a scenario
+            if((availableForceIDs.size() > 0) &&
+                    (Compute.randomInt(100) <= track.getScenarioOdds())) {
+                // pick random coordinates and force to drive the scenario
+                int x = Compute.randomInt(track.getWidth());
+                int y = Compute.randomInt(track.getHeight());                
+                
+                StratconCoords scenarioCoords = new StratconCoords(x, y);
+                
+                int randomForceIndex = Compute.randomInt(availableForceIDs.size());
+                int randomForceID = availableForceIDs.get(randomForceIndex);
+                
+                // remove the force from the available lists so we don't designate it as primary twice
+                availableForceIDs.remove(randomForceIndex); 
+                
+                // we want to remove the actual int with the value, not the value at the index
+                sortedAvailableForceIDs.get(MapLocation.AllGroundTerrain).remove((Integer) randomForceID);
+                sortedAvailableForceIDs.get(MapLocation.LowAtmosphere).remove((Integer) randomForceID);
+                sortedAvailableForceIDs.get(MapLocation.Space).remove((Integer) randomForceID);
+                
+                // two scenarios on the same coordinates wind up increasing in size
+                if(track.getScenarios().containsKey(scenarioCoords)) {
+                    track.getScenarios().get(scenarioCoords).incrementRequiredPlayerLances();
+                    assignAppropriateExtraForceToScenario(track.getScenarios().get(scenarioCoords), sortedAvailableForceIDs, campaign);
+                    continue;
+                }
+                
+                StratconScenario scenario = setupScenario(scenarioCoords, randomForceID, campaign, contract, track);
+                generatedScenarios.add(scenario);
+                
+                // if we're auto-assigning lances, deploy the force to the track as well
+                if(autoAssignLances) {
+                    processForceDeployment(scenarioCoords, randomForceID, campaign, track);
+                }
+            }
+        }
+        
+        // if under liaison command, pick a random scenario from the ones generated
+        // to set as required and attach liaison
+        if(contract.getCommandRights() == AtBContract.COM_LIAISON) {
+            int scenarioIndex = Compute.randomInt(generatedScenarios.size() - 1);
```

Comment:
Why -1? Is the last scenario not going to have liaison units?

Reply:
Replaced with Utilities.getRandomItem()